File: othello-client/src/othello.c
#include <stdio.h>
#include <memory.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "othello.h"

void initialize_board(char b[OTHELLO_ROW][OTHELLO_COL]){
    int i,j;
    for(i=0;i<OTHELLO_ROW;i++){
        for(j=0;j<OTHELLO_COL;j++){
            b[i][j]=' ';
        }
    }
    b[3][3]=BLACK;
    b[4][4]=BLACK;
    b[3][4]=WHITE;
    b[4][3]=WHITE;
}
void print_board(char b[OTHELLO_ROW][OTHELLO_COL]){
    int i,j;
    printf(" ");
    for(i=0;i<OTHELLO_COL;i++){
        printf(" %d",i);
    }
    printf("\n");
    printf(" ");
    for(i=0;i<OTHELLO_COL*2+1;i++){
        printf("-");
    }
    printf("\n");
    for(i=0;i<OTHELLO_ROW;i++){
        printf("%d",i);
        for(j=0;j<OTHELLO_COL;j++){
            printf("|%c",b[i][j]);
        }
        printf("|\n");
    }
    printf(" ");
    for(i=0;i<OTHELLO_COL*2+1;i++){
        printf("-");
    }
    printf("\n");
}

int count_flip_stone(char turn, int row, int col, int d_row, int d_col, char b[OTHELLO_ROW][OTHELLO_COL]){
    if(row+d_row<0 || col+d_col<0){
        return 0;
    }
    if(row+d_row>7 || col+d_col>7){
        return 0;
    }
    if(b[row+d_row][col+d_col]==turn){
        return 0;
    }
    int t_row=row, t_col=col;
    int flip=0;
    int find_mystone=0;
    while(!(t_row>7) || !(t_col>7)||!(t_row<0)||!(t_col<0)){
        t_row=t_row+d_row;
        t_col=t_col+d_col;
        if(b[t_row][t_col]==' '){
            break;
        }
        if(b[t_row][t_col]==turn){
            find_mystone=1;
            break;
        }
        flip++;
    }
    return (find_mystone==1)?flip:0;
}
void flip_stone(char turn, int row, int col, int d_row, int d_col, char b[OTHELLO_ROW][OTHELLO_COL]){
    int t_row=row, t_col=col;
    int flip=0;
    int find_mystone=0;
    while(!(t_row>7) || !(t_col>7)||!(t_row<0)||!(t_col<0)){
        t_row=t_row+d_row;
        t_col=t_col+d_col;
        if(b[t_row][t_col]==turn){
            break;
        }
        b[t_row][t_col]=turn;
    }
}
int flip(char turn, int row, int col, char b[OTHELLO_ROW][OTHELLO_COL]){
    b[row][col]=turn;
    int count=0;
    int d_row=-1, d_col=-1;
    for(d_row=-1;d_row<=1;d_row++){
        for(d_col=-1;d_col<=1;d_col++){
            count=count_flip_stone(turn,row,col,d_row,d_col,b);
            if(count>0){
                flip_stone(turn,row,col,d_row,d_col,b);
            }
        }
    }
    return count;
}
int check_position(char turn, int row, int col, char b[OTHELLO_ROW][OTHELLO_COL]){
    if(b[row][col]!=NO_STONE){
        return 0;
    }
    /*
    [-1,-1]左上　[-1, 0]上
    [-1, 1]右上　[ 0, 1]右
    [ 1, 1]右下　[ 1, 0]下
    [ 1,-1]左下　[ 0,-1]左
    */
    int count=0;
    int d_row=-1, d_col=-1;
    for(d_row=-1;d_row<=1;d_row++){
        for(d_col=-1;d_col<=1;d_col++){
            count+=count_flip_stone(turn,row,col,d_row,d_col,b);
        }
    }
    return count;
}
int make_connection(char* addr, int *sock){
    struct in_addr dest;
    inet_aton(addr, &dest);
    struct sockaddr_in srv_addr;

    // 接続先アドレスを設定する
    memcpy(&(srv_addr.sin_addr.s_addr),(char*)&(dest.s_addr),4);
    srv_addr.sin_port=htons(PORT);
    srv_addr.sin_family=AF_INET;

    *sock = socket(AF_INET, SOCK_STREAM, 0);
    
    
    if (*sock < 0) {
        perror("ERROR opening socket");
        exit(1);
    }

    // サーバーのホスト名を取得する
    /*
    struct hostent *server;
    server = gethostname(addr);
    if (server == NULL) {
        fprintf(stderr, "ERROR, no such host\n");
        exit(0);
    }
    */

    // サーバーに接続する
    if (connect(*sock, (struct sockaddr *)&srv_addr, sizeof(srv_addr)) < 0) {
        perror("ERROR connecting");
        exit(1);
    }
}

char initialize_game(int sock, char *user_name){
    int n=0;
    msg m;
    memset(&m,0,MESSAGE_LENGTH);
    memcpy(m.name,user_name,strlen(user_name));
    m.type=CONN_REQ;
    /*
    msg.type: CONN_REQ
    msg.name: my user name
    msg.color: N/A
    msg.row: N/A
    msg.col: N/A
    */
    n=send(sock, &m, MESSAGE_LENGTH, 0);
    if (n < 0) {
        perror("ERROR sending from socket");
        exit(1);
    }
    memset(&m,0,MESSAGE_LENGTH);
    n=recv(sock, &m, MESSAGE_LENGTH, 0);
    if (n < 0) {
        perror("ERROR reading from socket");
        exit(1);
    }
    if(m.type!=CONN_RES){
        printf("initialize process is not done.\n");
        exit(1);
    }
    /*
    msg.type: CONN_RES
    msg.name: a name of opponent
    msg.color: my color
    msg.row: N/A
    msg.col: N/A
    */
    printf("%s vs %s\n",user_name,m.name);
    printf("My color is %c\n",(m.color));
    return m.color;
    
}
int main(int argc, char** argv){
    const char* optstring = "s:n:";
    int c;
    char *srv_addr=NULL;
    char *user_name=NULL;

    if(argc!=5){
        printf("args:%d\n",argc);
        printf(" arg -s server_address, -n user_name\n");
        exit(0);
    }
    while ((c=getopt(argc, argv, optstring)) != -1) {
        //printf("opt=%c ", c);
        switch(c){
            case 's':{
                srv_addr=optarg;
                break;
            }
            case 'n':{
                user_name=optarg;
                break;
            }
            default:{
                printf(" arg -s server_address, -n user_name\n");
                exit(0);
            }
        }
    }
    printf("server address:%s\n",srv_addr);
    if(strlen(user_name)>20){
        printf("the length of the input user name must be less than 20 characters\n.");
        exit(0);
    }
    printf("user name:%s\n",user_name);

    int sock;
    make_connection(srv_addr,&sock);// make socket communicating with a server by using an IP address
    char mycolor=initialize_game(sock,user_name);// sending a message to connect the server and receive my color

    char board[OTHELLO_ROW][OTHELLO_COL];
    initialize_board(board);
    

    char turn=BLACK;
    int row=0,col=0,rcv_cnt;
    msg m;
    while(1){
        memset(&m,0,MESSAGE_LENGTH);
        print_board(board);
        if(turn!=mycolor){
            printf("--- waiting for my turn ---\n");
            rcv_cnt=recv(sock,&m,MESSAGE_LENGTH,0);
            /*
            msg.type: PUT_OPP_STONE
            msg.name: a name of opponent
            msg.color: opponent color
            msg.row: a row index of the opponent stone
            msg.col: a col index of the opponent stone
            */
            if(rcv_cnt<0){
                perror("ERROR reading a message including an opponet stone from socket");
                exit(1);
            }
            if(m.type!=PUT_OPP_STONE){
                printf("received message does not contain an opponent stone info\n");
                exit(1);
            }
            flip(turn,m.row,m.col,board);
            print_board(board);
            turn=mycolor;
        }
        printf("My turn \"%c\"\n",mycolor);
        printf("input row(0-7):");
        scanf("%d",&row);
        printf("input col(0-7):");
        scanf("%d",&col);
        printf("(row, col)=(%d,%d)\n",row,col);
        
        if(check_position(turn,row,col,board)==0){
            printf("you cannot put your stone at (%d, %d).\n",row,col);
            continue;
        }
        memset(&m,0,MESSAGE_LENGTH);
        m.type=PUT_MY_STONE;
        memcpy(m.name,user_name,NAME_LENGTH);
        m.color=mycolor;
        m.row=row;
        m.col=col;
        send(sock,&m,MESSAGE_LENGTH,0);
        flip(turn,row,col,board);

        turn=(turn==BLACK)?WHITE:BLACK;
    }
}

File: othello-client/src/othello.py
import socket
import struct
import pygame
import sys

OTHELLO_ROW = 8
OTHELLO_COL = 8
PORT = 10000 
MESSAGE_LENGTH = 64  # 仮設定、必要に応じて調整
NAME_LENGTH = 20
NO_STONE = ' '
EMPTY = 0
BLACK = 'B'
WHITE = 'W'
CONN_REQ = 1
CONN_RES = 2
PUT_MY_STONE = 3
PUT_OPP_STONE = 4
WIDTH, HEIGHT = 640, 640

class Msg:
    def __init__(self):
        self.type = 0
        self.name = ''
        self.color = ''
        self.row = -1
        self.col = -1

    def serialize(self):
        name_bytes = self.name.encode('utf-8')[:NAME_LENGTH]
        name_bytes += b'\x00' * (NAME_LENGTH - len(name_bytes))
        return struct.pack('!B20sccBB', self.type, name_bytes, self.color.encode(), b'\x00', self.row, self.col)

    @classmethod
    def deserialize(cls, data):
        m = cls()
        m.type, name_bytes, color, _, m.row, m.col = struct.unpack('!B20sccBB', data)
        m.name = name_bytes.decode('utf-8').strip('\x00')
        m.color = color.decode('utf-8')
        return m

def initialize_board():
    board = [[NO_STONE for _ in range(OTHELLO_COL)] for _ in range(OTHELLO_ROW)]
    board[3][3] = BLACK
    board[4][4] = BLACK
    board[3][4] = WHITE
    board[4][3] = WHITE
    return board

def print_board(board):
    print('  ' + ' '.join(map(str, range(OTHELLO_COL))))
    print(' ' + '-' * (OTHELLO_COL * 2 + 1))
    for i, row in enumerate(board):
        print(str(i) + ''.join(f'|{cell}' for cell in row) + '|')
    print(' ' + '-' * (OTHELLO_COL * 2 + 1))

def count_flip_stone(turn, row, col, d_row, d_col, board):
    t_row, t_col = row + d_row, col + d_col
    if not (0 <= t_row < OTHELLO_ROW and 0 <= t_col < OTHELLO_COL):
        return 0
    if board[t_row][t_col] == turn or board[t_row][t_col] == NO_STONE:
        return 0
    flip = 0
    while 0 <= t_row < OTHELLO_ROW and 0 <= t_col < OTHELLO_COL:
        if board[t_row][t_col] == NO_STONE:
            return 0
        if board[t_row][t_col] == turn:
            return flip
        flip += 1
        t_row += d_row
        t_col += d_col
    return 0

def flip_stone(turn, row, col, d_row, d_col, board):
    t_row, t_col = row + d_row, col + d_col
    while 0 <= t_row < OTHELLO_ROW and 0 <= t_col < OTHELLO_COL:
        if board[t_row][t_col] == turn:
            break
        board[t_row][t_col] = turn
        t_row += d_row
        t_col += d_col

def flip(turn, row, col, board):
    board[row][col] = turn
    for d_row in (-1, 0, 1):
        for d_col in (-1, 0, 1):
            if d_row == 0 and d_col == 0:
                continue
            if count_flip_stone(turn, row, col, d_row, d_col, board) > 0:
                flip_stone(turn, row, col, d_row, d_col, board)

def check_position(turn, row, col, board):
    if board[row][col] != NO_STONE:
        return 0
    count = 0
    for d_row in (-1, 0, 1):
        for d_col in (-1, 0, 1):
            if d_row == 0 and d_col == 0:
                continue
            count += count_flip_stone(turn, row, col, d_row, d_col, board)
    return count

def make_connection(addr):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((addr, PORT))
    return sock

def initialize_game(sock, user_name):
    m = Msg()
    m.type = CONN_REQ
    m.name = user_name
    sock.sendall(m.serialize())
    data = sock.recv(MESSAGE_LENGTH)
    m = Msg.deserialize(data)
    if m.type != CONN_RES:
        print("initialize process is not done.")
        exit(1)
    print(f"{user_name} vs {m.name}")
    print(f"My color is {m.color}")
    return m.color




def draw_board(screen, board):
    cell_size = WIDTH // 8
    # 緑の背景
    screen.fill((0, 128, 0))
    # 盤の線
    for i in range(9):
        pygame.draw.line(screen, (0, 0, 0), (i*cell_size, 0), (i*cell_size, HEIGHT))
        pygame.draw.line(screen, (0, 0, 0), (0, i*cell_size), (WIDTH, i*cell_size))
    # 石を描く
    for row in range(8):
        for col in range(8):
            if board[row][col] != EMPTY:
                color = (0,0,0) if board[row][col]==BLACK else (255,255,255)
                center = (col*cell_size + cell_size//2, row*cell_size + cell_size//2)
                pygame.draw.circle(screen, color, center, cell_size//2 - 5)


def main():
    import argparse

    pygame.init()

    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption('Othello')
    EMPTY = 0
    BLACK = 1
    WHITE = 2

    board = [[EMPTY for _ in range(8)] for _ in range(8)]
    board[3][3] = WHITE
    board[4][4] = WHITE
    board[3][4] = BLACK
    board[4][3] = BLACK

    draw_board(screen, board)

    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
    parser = argparse.ArgumentParser()
    parser.add_argument('-s', required=True, help='Server address')
    parser.add_argument('-n', required=True, help='User name')
    args = parser.parse_args()

    srv_addr = args.s
    user_name = args.n

    if len(user_name) > 20:
        print("the length of the input user name must be less than 20 characters.")
        exit(0)

    print(f"server address: {srv_addr}")
    print(f"user name: {user_name}")

    sock = make_connection(srv_addr)
    mycolor = initialize_game(sock, user_name)
    board = initialize_board()

    turn = BLACK
    while True:
        print_board(board)
        if turn != mycolor:
            print("--- waiting for my turn ---")
            data = sock.recv(MESSAGE_LENGTH)
            m = Msg.deserialize(data)
            if m.type != PUT_OPP_STONE:
                print("received message does not contain an opponent stone info")
                exit(1)
            flip(turn, m.row, m.col, board)
            turn = mycolor
            continue

        print(f"My turn \"{mycolor}\"")
        row = int(input("input row(0-7): "))
        col = int(input("input col(0-7): "))
        print(f"(row, col)=({row},{col})")

        if check_position(turn, row, col, board) == 0:
            print(f"you cannot put your stone at ({row}, {col}).")
            continue

        m = Msg()
        m.type = PUT_MY_STONE
        m.name = user_name
        m.color = mycolor
        m.row = row
        m.col = col
        sock.sendall(m.serialize())
        flip(turn, row, col, board)

        turn = WHITE if turn == BLACK else BLACK

if __name__ == "__main__":
    main()
